### ISSUE TITLE
Data race warning

### ISSUE THREAD

**Issue by Perlence** [franela/goreq#111]
I use goreq to send concurrent requests, but when I checked the program for races, I got this:

```
WARNING: DATA RACE
Write by goroutine 14:
  github.com/franela/goreq.Request.Do()
      /home/ubuntu/.local/go/src/github.com/franela/goreq/goreq.go:314 +0x6ff
```

So race detector warns me that goroutines write unsafely into `CheckRedirect` of `DefaultClient`.

**Comment by marcosnils** [comment]
@Perlence can you update goreq and run the race detector again?. Seems like is something else https://github.com/franela/goreq/blob/master/goreq.go#L314

**Comment by Perlence** [comment]
I've updated goreq so the master is 8f5efc2. Unfortunately warning's still there:

```
Write by goroutine 14:
  github.com/franela/goreq.Request.Do()
      /home/ubuntu/.local/go/src/github.com/franela/goreq/goreq.go:323 +0x6ff
```

**Comment by marcosnils** [comment]
@Perlence We found the issue. We'll be fixing this shorty.

**Comment by rhd** [comment]
Hi, I'm seeing the same issue - would love a fix just to clean up the warning.

**Comment by marcosnils** [comment]
@rhd we'll fix it shortly.

**Comment by rrh** [comment]
I have independently rediscovered this race condition in my application.  I'm at head, eg 72c51a544272e007ab3da4f7d9ac959b7af7af03

Note also that changing the Makefile to run go test --race finds other race conditions as well.

**Comment by sschepens** [comment]
Yes, there are some race conditions they all come from using DefaultClient and allowing functionallity that should require a new client.
I believe we should add a Client abstraction to goreq and make these things configurable at a global level and allow to create new goreq clients.

**Comment by rrh** [comment]
any progress on this over the weekend?

**Comment by sschepens** [comment]
@rrh I haven't had time to check this, but those races are hard to tackle without breaking backwards compatibility, that's a bummer.
